atoma-proxy/src/server/handlers/completions.rs

```diff
@@ -0,0 +1,1311 @@
+    COMPLETION_TOKENS, PROMPT_TOKENS, STOP_STREAMER_PATH, STREAM, TOTAL_TOKENS, USAGE, USER_ID,
+};
+use crate::server::{Result, DEFAULT_MAX_TOKENS, MAX_COMPLETION_TOKENS, MAX_TOKENS, MODEL};
+
+/// Path for the completions endpoint.
+///
+/// This endpoint follows the OpenAI API format for legacy completions
+/// and is used to process chat-based requests for AI model inference.
+pub const COMPLETIONS_PATH: &str = "/v1/completions";
+
+/// Path for the confidential completions endpoint.
+pub const CONFIDENTIAL_COMPLETIONS_PATH: &str = "/v1/completions";
+
+/// The key for the prompt in the request.
+const PROMPT: &str = "prompt";
+
+/// The OpenAPI schema for the completions endpoint.
+#[derive(OpenApi)]
+#[openapi(
+    paths(completions_create),
+    components(schemas(
+        CompletionsRequest,
+        CompletionsResponse,
+        LogProbs,
+        Usage,
+        CompletionTokensDetails,
+        PromptTokensDetails,
+    ))
+)]
+pub struct CompletionsOpenApi;
+
+/// Create completion
+///
+/// This function processes completion requests by using the chat completions endpoint.
+///
+/// ## Returns
+///
+/// Returns a Response containing either:
+/// - A streaming SSE connection for real-time completions
+/// - A single JSON response for non-streaming completions
+///
+/// ## Errors
+///
+/// Returns an error status code if:
+/// - The request processing fails
+/// - The streaming/non-streaming handlers encounter errors
+/// - The underlying inference service returns an error
+#[utoipa::path(
+    post,
+    path = "",
+    security(
+        ("bearerAuth" = [])
+    ),
+    request_body = CompletionsRequest,
+    responses(
+        (status = OK, description = "Chat completions", content(
+            (CompletionsResponse = "application/json"),
+        )),
+        (status = BAD_REQUEST, description = "Bad request"),
+        (status = UNAUTHORIZED, description = "Unauthorized"),
+        (status = INTERNAL_SERVER_ERROR, description = "Internal server error")
+    )
+)]
+#[instrument(
+    level = "info",
+    skip_all,
+    fields(
+        path = COMPLETIONS_PATH,
+    )
+)]
+pub async fn completions_create(
+    Extension(metadata): Extension<RequestMetadataExtension>,
+    State(state): State<ProxyState>,
+    headers: HeaderMap,
+    Json(payload): Json<Value>,
+) -> Result<Response<Body>> {
+    let endpoint = metadata.endpoint.clone();
+    tokio::spawn(async move {
+        let is_streaming = payload
+            .get(STREAM)
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or_default();
+
+        match handle_completions_request(&state, &metadata, headers, payload, is_streaming).await {
+            Ok(response) => {
+                if !is_streaming {
+                    // The streaming metric is recorded in the streamer (final chunk)
+                    TOTAL_COMPLETED_REQUESTS.add(1, &[KeyValue::new("model", metadata.model_name)]);
+                }
+                Ok(response)
```

Comment:
Update stack/Fiat. So the amount is unlocked and counted as completed.

Reply:
This is done within the function body for non-streaming requests, and handle final chunk or at drop for streaming requests.